catkin_ws/src/stability/src/AxialPid.cpp

```diff
@@ -0,0 +1,257 @@
+}
+
+//returns whether the current roll position is at the target (within a certain margin)
+bool RotPIDController::atRollTarget(float rollValue)
+{
+	rollVal = rollValue;
+	return (fabs(rollVal - rollTarget) <= MARGIN);
+}
+
+//returns whether the current pitch position is at the target (within a certain margin)
+bool RotPIDController::atPitchTarget(float pitchValue)
+{
+	pitchVal = pitchValue;
+	return (fabs(pitchVal - pitchTarget) <= MARGIN);
+}
+
+//returns whether the current yaw position is at the target (within a certain margin)
+bool RotPIDController::atYawTarget(float yawValue)
+{
+	yawVal = yawValue;
+	return (fabs(yawVal - yawTarget) <= MARGIN);
+}
+
+//returns whether the current rotation is at the target (within a certain margin)
+bool RotPIDController::atTarget(float rollValue, float pitchValue, float yawValue)
+{
+	rollVal = rollValue;
+	pitchVal = pitchValue;
+	yawVal = yawValue;
+	return atRollTarget(rollValue) && atPitchTarget(pitchValue) && atYawTarget(yawValue);
+}
+
+
+
+//flips pid control (for if the robot is moving away from target instead of towards it)
+void RotPIDController::flipRoll()
+{
+	rollFlip = !rollFlip;
+	rollControl.setDirection(rollFlip);
+}
+
+void RotPIDController::flipPitch()
+{
+	pitchFlip = !pitchFlip;
+	pitchControl.setDirection(pitchFlip);
+}
+
+void RotPIDController::flipYaw()
+{
+	yawFlip = !yawFlip;
+	yawControl.setDirection(yawFlip);
+}
+
+//set pid tunings
+void RotPIDController::setRollPID(float kp, float ki, float kd)
+{
+	rollKp = kp;
+	rollKi = ki;
+	rollKd = kd;
+
+	rollControl.setPID(kp, ki, kd);
+}
+void RotPIDController::setPitchPID(float kp, float ki, float kd)
+{
+	pitchKp = kp;
+	pitchKi = ki;
+	pitchKd = kd;
+
+	pitchControl.setPID(kp, ki, kd);
+}
+void RotPIDController::setYawPID(float kp, float ki, float kd)
+{
+	yawKp = kp;
+	yawKi = ki;
+	yawKd = kd;
+
+	yawControl.setPID(kp, ki, kd);
+}
+
+//set pid target values
+void RotPIDController::setRollTarget(float newTarget)
+{
+	rollTarget = newTarget;
+	//rollControl.setSetpoint(newTarget);
+}
+
+void RotPIDController::setPitchTarget(float newTarget)
+{
+	pitchTarget = newTarget;
+	//pitchControl.setSetpoint(newTarget);
+}
+
+void RotPIDController::setYawTarget(float newTarget)
+{
+	yawTarget = newTarget;
+	//yawControl.setSetpoint(newTarget);
+}
+
+void RotPIDController::setTargets(geometry_msgs::Vector3 input)
+{
+	pitchTarget = input.y;
+	rollTarget = input.x;
+	yawTarget = input.z;
+}
+
+void RotPIDController::setInputs(sensor_msgs::Imu input)
+{
+  	EulerAngles angles = ToEulerAngles(input);
+  	pitchVal = angles.pitch;
+  	yawVal = angles.yaw;
+  	rollVal = angles.roll;	
+}
+
+
+void RotPIDController::runSubs(int argc, char** argv)
+{
+	ros::init(argc, argv, "pidLoop");
+	ros::NodeHandle n;
+	target = n.subscribe("target_topic", 50, &RotPIDController::setTargets, this);
+	sensor = n.subscribe("/aquadrone_v2/out/imu", 50, &RotPIDController::setInputs, this);	
+	ros::spinOnce();
+}
+
+
+//copied from wikipedia

```

Comment:
Can you add the URL of the page and/or heading that you got the formula from? There are different choices for Euler angles, so we want to be able to track which choice lead to these calculations.

Reply:
url was inputted